Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -1262,7 +1269,6 @@ WebGLRenderingContextBase::~WebGLRenderingContextBase()
         detachAndRemoveAllObjects();
         loseExtensions(LostContextMode::RealLostContext);
         destroyGraphicsContextGL();
-        m_contextGroup->removeContext(*this);
```

Comment:
why are you moving this to the beginning of the function instead of moving it just before the previous line where the garphicsContexGL is destroyed? Wouldn't that work for some reason that I'm missing?

Reply:
Hello @magomez ,

This context removal from group ('m_contextGroup->removeContext(*this)') must happen before the destructor of given WebGLSharedObject is called. Otherwise what happens in ~WebGLRenderingContextBase is like this:

- first, the reference is nullified ('m_boundArrayBuffer = nullptr' etc.); these are RefPtr's and this decreases the refcnt to this WebGLSharedObject to 0
- so, the destructor for this WebGLSharedObject is executed - this goes via WebGLObject::runDestructor() -> WebGLObject::deleteObject. WebGLObject::deleteObject checks for attachement count (m_attachmentCount) - and if it's not 0 it will not execute deleteObjectImpl (where the actual 'physical' removal from the context takes place). So the reference to the object is destructed, and ~WebGLSharedObject also removes the object from m_contextGroup. At this point, the resource is leaked I believe.

So only moving 'm_contextGroup->removeContext(*this)' before destroyGraphicsContextGL() isn't enough; it needs to happen before these destructors execute. With the submitted fix, m_contextGroup->removeContext(*this) is executed at the beginning of ~WebGLRenderingContextBase - and this calls detach() on everything (WebGLContextGroup::detachAndRemoveAllObjects) so is possible to release all the objects.

Please let me know if that makes sense now (suppose I could explain this better in the commit message?)

Regards,
Tomasz

Reply:
Ok, that makes sense. Please split this commit into 2 different ones, because its doing two different things: one fixes the leak and the other is a trick to release memory faster, but it's not related. After that it should be ok to merge.